### ISSUE TITLE
Incorrect VMOVAPD code generation

### ISSUE THREAD

**Issue by gwoltman** [Terraspace/UASM#70]
In Uasm 24.6.4

`vmovapd zmm10, zmm17`

assembles as

`        vmovapd zmm10, zmm1                             ; 0000 _ 62 71 FD 48: 28. D1`

**Comment by john-terraspace** [comment]
Hi,

 

This has been fixed as well as the kmov errors. I’ve added these as two new sets of regression tests in 2.46.5.

 

I’ve also added a set of regression tests to look at evex use with equates and macros relating to your other report, which is as follows:

 

 

KMASK equ <k3>

 

TITLE setup

OPTION EVEX:1

OPTION FLAT:1

 

zfmsubpd MACRO reg, mask, reg2, reg3, reg4

ENDM

 

.code

USE64

 

;--------------------------------------------------------

 

vsubpd zmm0 {KMASK}, zmm1, zmm2 ; direct macro substitution works.

 

;--------------------------------------------------------

 

znoconst_rounding_single_nottp MACRO zmmval, zmmcarry, kwritemask

zfmsubpd zmmcarry, {kwritemask}, zmmval, zmm24, zmm20

zfmsubpd zmm16, {kwritemask}, zmmcarry, zmm26, zmm22

vsubpd zmmval {kwritemask}, zmmval, zmm16

ENDM

 

;--------------------------------------------------------

 

znoconst MACRO zmmval, kwritemask

 

 

vsubpd zmmval {kwritemask}, zmmval, zmm16

                

; multiple elements can be substituted in the macro, include 

; placeholder for register and decorator.

 

vsubpd zmm2 {kwritemask}, zmmval, zmm16

 

; register + expanded decorator from macro parameter works.

 

ENDM

 

;--------------------------------------------------------

 

znoconst_rounding_single_nottp zmm3, zmm4, k6

znoconst zmm2, k1

 

;0:  62 f1 f5 4b 5c c2       vsubpd zmm0{k3},zmm1,zmm2

;6:  62 b1 e5 4e 5c d8       vsubpd zmm3{k6},zmm3,zmm16

;c:  62 b1 ed 49 5c d0       vsubpd zmm2{k1},zmm2,zmm16

;12: 62 b1 ed 49 5c d0       vsubpd zmm2{k1},zmm2,zmm16

 

This is working as expected now.

With regards to your previous macro, we realized that you cannot pass a reg {decorator} as a single parameter, that will never work as they’re totally separate items, so I modified the macro to take the register and the write mask as separate arguments. There was a bug with how equate values were only being processed after evex decorators which is now fixed as well allowing things like:

 

vsubpd zmm0 {KMASK}, zmm1, zmm2 ; direct macro substitution works.

 

 

This will all be up tonight in 2.46.5 update.

 

Thanks for all the reports! 😊

 

John
